### PR TITLE
[WP] Refactor custom events to properly send them

### PR DESCRIPTION
**Pull request by theop-dd**
### What does this PR do?
- refactor a bit how custom events are handled so we always send them each time an action report is sent.
- add a new status to the kill action to handle early exit processes. When we try to kill one process (scope process), if the process exits before we kill it, we abort the kill instead of sending `kill_queued` status.

Note: This PR only fixes issues for scope process. Other changes will be needed in the future to handle other hedge cases.

### Motivation
Now every time an action report is sent, we have an associated remediation status event sent. Previously, we had situation where one was emitted but not the second one.

### CODE REVIEW

File: pkg/security/module/server.go [hunk 1]
```diff
@@ -359,6 +368,8 @@ func (a *APIServer) start(ctx context.Context) {
 					return false
 				}
 
+				// For kill actions, send a custom remediation event per resolved kill report
+				SendCustomEventKillAction(a.probe, msg.tags, msg.actionReports)
```
Comment (spikat) [inline pkg/security/module/server.go:373]: Here we are sending an event for ALL kill actions. What if a rule has 2 rules action, a first one with disabled disarmers (so, killing at trigger) and another one with a delayed kill by disarmers? The first performed kill will be sent another time?

File: pkg/security/probe/actions.go [hunk 2]
```diff
@@ -55,9 +57,11 @@ type KillActionReport struct {
...
 
 	// internal
-	Pid      uint32
-	resolved bool
-	rule     *rules.Rule
+	Pid                uint32
+	resolved           bool
+	rule               *rules.Rule
+	containerID        string
+	containerCreatedAt uint64
```
Comment (spikat) [inline pkg/security/probe/actions.go]: Please add comments here, to remember is needed for remediation status events and would not be serialized on the killactionreport, and that the containerID could be nil

Comment (spikat) [reply pkg/security/probe/actions.go]: plus: could you use directly the RemediationContainerContext here?

File: pkg/security/probe/actions.go [hunk 3]
```diff
@@ -128,3 +132,10 @@ func (k *KillActionReport) IsMatchingRule(ruleID eval.RuleID) bool {
 
 	return k.rule.ID == ruleID
 }
+
+// GetContainerContext returns the container ID and created_at when the process was in a container, otherwise empty/zero.
+func (k *KillActionReport) GetContainerContext() (containerID string, createdAt uint64) {
```
Comment (spikat) [inline pkg/security/probe/actions.go]: could you return directly a RemediationContainerContext?

File: pkg/security/probe/process_killer.go [hunk 4]
```diff
@@ -241,6 +241,14 @@ func (p *ProcessKiller) HandleProcessExited(event *model.Event) {
 		defer report.Unlock()
 
 		if report.Pid == event.ProcessContext.Pid {
+			if report.Scope == "process" {
+				if report.Status == KillActionStatusQueued {
+					// The process exited before the kill was performed
+					report.Status = KillActionStatusKillAborted
+				}
+				report.resolved = true
+				return true
```
Comment (spikat) [inline pkg/security/probe/process_killer.go]: You shall remove those 2 lines

File: pkg/security/tests/action_test.go [hunk 5]
```diff
@@ -1831,3 +1831,178 @@ func TestRemediationCustomEventNotTriggered(t *testing.T) {
...
+	// Reload the policy and wait for the kill rule to trigger
+	// The rule should match the replayed exec event for the container's entrypoint (cat)
+	err = test.reloadPolicies()
+	if err != nil {
+		t.Error("failed to reload policies: %w", err)
+	}
+	// Trigger a small event to force the replay of cached events.
+	// The replay only happens in handleEvent when a new eBPF event arrives.
+	time.Sleep(2500 * time.Millisecond)
+	exec.Command("true").Run()
```
Comment (spikat) [inline pkg/security/tests/action_test.go]: this part should not be needed anymore

File: pkg/security/tests/action_test.go [hunk 6]
```diff
@@ -1831,3 +1831,178 @@ func TestRemediationCustomEventNotTriggered(t *testing.T) {
...
+		})
+
+		return nil
+	}, retry.Delay(200*time.Millisecond), retry.Attempts(30), retry.DelayType(retry.FixedDelay))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Verify that the container was killed by checking if it's still running
+	err = retry.Do(func() error {
```
Comment (spikat) [inline pkg/security/tests/action_test.go:1991]: not sure we need a retry here

File: pkg/security/module/server.go (hunk 1 above)
Comment (spikat) [reply pkg/security/module/server.go:373]: This corner case could be handled in the second rework PR, but we shouldn't forget to do it. Meanwhile please add a comment to not forgot

File: pkg/security/probe/process_killer.go (hunk 4 above)
Comment (theop-dd) [reply pkg/security/probe/process_killer.go]: If you want I can include it in the scope process but if I remove it we won't send the event when the process exit (it's useless to wait).
We know that's not optimal but it should be part of the big refacto imho

File: pkg/security/tests/action_test.go (hunk 5 above)
Comment (theop-dd) [reply pkg/security/tests/action_test.go]: Nop :) If I remove it the test still fails.

File: pkg/security/probe/actions.go (hunk 3 above)
Comment (theop-dd) [reply pkg/security/probe/actions.go]: yes!

File: pkg/security/module/server.go (hunk 1 above)
Comment (theop-dd) [reply pkg/security/module/server.go:373]: On the same process so? 
Super strange behavior indeed (good catch of another hedge case!).
We send one action report but we get 2 remediation status with `performed` :/

File: pkg/security/probe/actions.go (hunk 2 above)
Comment (theop-dd) [reply pkg/security/probe/actions.go]: yes!

File: pkg/security/module/server.go (hunk 1 above)
Comment (spikat) [reply pkg/security/module/server.go:373]: `We send one action report` not exactly, we sent ONE event, with TWO action report on it :)
And it could be the same process yes, like the first kill without disarmers on scope process, and the second one on scope cgroup, something like that

File: pkg/security/tests/action_test.go (hunk 5 above)
Comment (spikat) [reply pkg/security/tests/action_test.go]: arf :/ ok. the exec should be needed, I though the sleep neither but .. ok, let keep the sleep

File: pkg/security/module/server.go (hunk 1 above)
Comment (theop-dd) [reply pkg/security/module/server.go:373]: Yes true, one event with 2 reports. But 2 remediation events tho ;)

File: pkg/security/probe/process_killer.go (hunk 4 above)
Comment (spikat) [reply pkg/security/probe/process_killer.go]: lines removed, thanks

**Review by spikat — APPROVED** [review]
